test/Microsoft.NET.Build.Tests/RoslynBuildTaskTests.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+using Basic.CompilerLog.Util;
+using Microsoft.Build.Logging.StructuredLogger;
+
+namespace Microsoft.NET.Build.Tests;
+
+public sealed class RoslynBuildTaskTests(ITestOutputHelper log) : SdkTest(log)
+{
+    private const string CoreCompilerFileName = "csc.dll";
+    private const string FxCompilerFileName = "csc.exe";
+
+    [FullMSBuildOnlyTheory, CombinatorialData]
+    public void FullMSBuild_SdkStyle(bool useSharedCompilation)
+    {
+        var testAsset = CreateProject(useSharedCompilation);
+        var buildCommand = BuildAndRunUsingMSBuild(testAsset);
+        VerifyCompiler(buildCommand, CoreCompilerFileName, useSharedCompilation);
+    }
+
+    [FullMSBuildOnlyTheory, CombinatorialData]
+    public void FullMSBuild_SdkStyle_OptOut(bool useSharedCompilation)
+    {
+        var testAsset = CreateProject(useSharedCompilation).WithProjectChanges(static doc =>
+        {
+            doc.Root!.Element("PropertyGroup")!.Add(new XElement("RoslynCompilerType", "Framework"));
+        });
+        var buildCommand = BuildAndRunUsingMSBuild(testAsset);
+        VerifyCompiler(buildCommand, FxCompilerFileName, useSharedCompilation);
+    }
+
+    [FullMSBuildOnlyTheory, CombinatorialData]
+    public void FullMSBuild_NonSdkStyle(bool useSharedCompilation)
+    {
+        var testAsset = CreateProject(useSharedCompilation, static project =>
+        {
+            project.IsSdkProject = false;
+            project.TargetFrameworkVersion = "v4.7.2";
+        });
+        var buildCommand = BuildAndRunUsingMSBuild(testAsset);
+        VerifyCompiler(buildCommand, FxCompilerFileName, useSharedCompilation);
+    }
+
+    [Theory, CombinatorialData]
+    public void DotNet(bool useSharedCompilation)
+    {
+        var testAsset = CreateProject(useSharedCompilation);
+        var buildCommand = BuildAndRunUsingDotNet(testAsset);
+        VerifyCompiler(buildCommand, CoreCompilerFileName, useSharedCompilation);
+    }
+
+    private TestAsset CreateProject(bool useSharedCompilation, Action<TestProject>? configure = null, [CallerMemberName] string callingMethod = "")
+    {
+        var project = new TestProject
+        {
+            Name = "App1",
+            IsExe = true,
+            SourceFiles =
+            {
+                ["Program.cs"] = """
+                    class Program
+                    {
+                        static void Main()
+                        {
+                            System.Console.WriteLine(40 + 2);
+                        }
+                    }
+                    """,
+            },
+        };
+
+        // UseSharedCompilation should be the default, so set it only if false.
+        if (!useSharedCompilation)
+        {
+            project.AdditionalProperties["UseSharedCompilation"] = "false";
+        }
```

Comment:
Might be simpler to just always be explicit here:

```suggestion
        project.AdditionalProperties["UseSharedCompilation"] = useSharedCompilation ? "true" : "false";
        
```

Reply:
Which you can probably fold into the initializer above actually